File: src/priv/bajdcc/util/lexer/regex/RegexStringIterator.java
package priv.bajdcc.util.lexer.regex;

import java.util.ArrayList;
import java.util.Stack;

import priv.bajdcc.util.Position;
import priv.bajdcc.util.lexer.error.RegexException;
import priv.bajdcc.util.lexer.error.RegexException.RegexError;
import priv.bajdcc.util.lexer.token.MetaType;
import priv.bajdcc.util.lexer.token.Token;

/**
 * 字符串迭代器，提供字节流解析功能
 * 
 * @author bajdcc
 *
 */
public class RegexStringIterator implements IRegexStringIterator, Cloneable {

	/**
	 * 存储字符串
	 */
	protected String context;

	/**
	 * 用于恢复的位置堆栈
	 */
	public Stack<Integer> stkIndex = new Stack<Integer>();

	/**
	 * 位置
	 */
	protected Position position = new Position();

	/**
	 * 用于恢复行列数的堆栈
	 */
	public Stack<Position> stkPosition = new Stack<Position>();

	/**
	 * 当前的分析信息
	 */
	protected RegexStringIteratorData data = new RegexStringIteratorData();
	
	/**
	 * 记录每行起始的位置
	 */
	protected ArrayList<Integer> arrLinesNo = new ArrayList<Integer>();

	public RegexStringIterator() {
		arrLinesNo.add(0);
	}

	public RegexStringIterator(String context) {
		this();
		this.context = context;
	}

	/**
	 * 字符解析组件
	 */
	protected RegexStringUtility utility = new RegexStringUtility(this);

	@Override
	public void err(RegexError error) throws RegexException {
		throw new RegexException(error, position);
	}

	@Override
	public void next() {
		if (available()) {
			advance();
		}
		translate();
		position.iColumn++;
		if (data.chCurrent == MetaType.NEW_LINE.getChar()) {
			arrLinesNo.add(arrLinesNo.get(arrLinesNo.size() - 1) + position.iColumn + 1);
			position.iColumn = 0;
			position.iLine++;
		}
	}

	@Override
	public Token scan() {
		return null;
	}

	@Override
	public Position position() {
		return position;
	}

	@Override
	public void translate() {
		if (!available()) {
			data.chCurrent = 0;
			data.kMeta = MetaType.END;
			return;
		}
		data.chCurrent = current();
		transform();
	}

	/**
	 * 分析字符类型
	 */
	protected void transform() {
		data.kMeta = MetaType.CHARACTER;
	}

	@Override
	public boolean available() {
		return data.iIndex >= 0 && data.iIndex < context.length();
	}

	@Override
	public void advance() {
		data.iIndex++;
	}

	@Override
	public char current() {
		return context.charAt(data.iIndex);
	}

	@Override
	public MetaType meta() {
		return data.kMeta;
	}

	@Override
	public int index() {
		return data.iIndex;
	}

	@Override
	public void expect(MetaType meta, RegexError error) throws RegexException {
		if (data.kMeta == meta) {
			next();
		} else {
			err(error);
		}
	}

	@Override
	public void snapshot() {
		stkIndex.push(data.iIndex);
		stkPosition.push(new Position(position.iColumn,
				position.iLine));
	}

	@Override
	public void cover() {
		stkIndex.set(stkIndex.size() - 1, data.iIndex);
		stkPosition.set(stkPosition.size() - 1, new Position(position));
	}

	@Override
	public void restore() {
		data.iIndex = stkIndex.pop();
		position = new Position(stkPosition.pop());
	}

	@Override
	public void discard() {
		stkIndex.pop();
		stkPosition.pop();
	}

	@Override
	public RegexStringUtility utility() {
		return utility;
	}

	@Override
	public String getRegexDescription() {
		return context;
	}

	@Override
	public IRegexStringIterator copy() {
		return null;
	}

	@Override
	protected Object clone() throws CloneNotSupportedException {
		RegexStringIterator o = (RegexStringIterator) super.clone();
		o.position = (Position) o.position.clone();
		o.data = (RegexStringIteratorData) o.data.clone();
		return o;
	}

	@Override
	public IRegexStringIteratorEx ex() {
		return null;
	}
}

File: src/main/java/com/bajdcc/util/lexer/regex/RegexStringIterator.kt
package com.bajdcc.util.lexer.regex

import com.bajdcc.util.Position
import com.bajdcc.util.lexer.error.RegexException
import com.bajdcc.util.lexer.error.RegexException.RegexError
import com.bajdcc.util.lexer.token.MetaType
import java.util.*

/**
 * 字符串迭代器，提供字节流解析功能
 *
 * @author bajdcc
 */
open class RegexStringIterator() : IRegexStringIterator, Cloneable {

    override fun regexDescription() = context

    /**
     * 存储字符串
     */
    open var context: String = ""

    /**
     * 用于恢复的位置堆栈
     */
    var stackIndex = Stack<Int>()

    /**
     * 位置
     */
    protected var position = Position()

    /**
     * 用于恢复行列数的堆栈
     */
    var stackPosition = Stack<Position>()

    /**
     * 当前的分析信息
     */
    protected var data = RegexStringIteratorData()

    /**
     * 记录每行起始的位置
     */
    protected var arrLinesNo = mutableListOf<Int>()

    /**
     * 字符解析组件
     */
    protected var utility = RegexStringUtility(this)

    init {
        arrLinesNo.add(-1)
    }

    constructor(context: String) : this() {
        this.context = context
    }

    @Throws(RegexException::class)
    override fun err(error: RegexError) {
        throw RegexException(error, position)
    }

    override fun next() {
        if (available()) {
            advance()
        }
        translate()
        position.column = position.column + 1
        if (data.current == MetaType.NEW_LINE.char) {
            val prev = arrLinesNo[arrLinesNo.size - 1]
            if (prev < data.index)
                arrLinesNo.add(data.index)
            position.column = 0
            position.line = position.line + 1
        }
    }

    override fun scan() {
        throw NotImplementedError()
    }

    override fun position(): Position {
        return position
    }

    override fun translate() {
        if (!available()) {
            data.current = '\u0000'
            data.meta = MetaType.END
            return
        }
        data.current = current()
        transform()
    }

    /**
     * 分析字符类型
     */
    protected open fun transform() {
        data.meta = MetaType.CHARACTER
    }

    override fun available(): Boolean {
        return data.index >= 0 && data.index < context.length
    }

    override fun advance() {
        data.index++
    }

    override fun current(): Char {
        return context[data.index]
    }

    override fun meta(): MetaType {
        return data.meta
    }

    override fun index(): Int {
        return data.index
    }

    @Throws(RegexException::class)
    override fun expect(meta: MetaType, error: RegexError) {
        if (data.meta === meta) {
            next()
        } else {
            err(error)
        }
    }

    override fun snapshot() {
        stackIndex.push(data.index)
        stackPosition.push(Position(position.column, position.line))
    }

    override fun cover() {
        stackIndex[stackIndex.size - 1] = data.index
        stackPosition[stackPosition.size - 1] = Position(position)
    }

    override fun restore() {
        data.index = stackIndex.pop()
        position = Position(stackPosition.pop())
    }

    override fun discard() {
        stackIndex.pop()
        stackPosition.pop()
    }

    override fun utility(): RegexStringUtility {
        return utility
    }

    override fun copy(): IRegexStringIterator {
        throw NotImplementedError()
    }

    override fun clone(): Any {
        val o = super.clone() as RegexStringIterator
        o.position = o.position.clone() as Position
        o.data = o.data.clone()
        o.utility = RegexStringUtility(o)
        return o
    }

    override fun ex(): IRegexStringIteratorEx {
        throw NotImplementedError()
    }
}